core/client/hdfs/src/main/java/alluxio/hadoop/LocalCachingFileSystem.java

```diff
@@ -0,0 +1,173 @@
+ */
+public class LocalCachingFileSystem extends org.apache.hadoop.fs.FileSystem {
+  private static final Logger LOG = LoggerFactory.getLogger(LocalCachingFileSystem.class);
+  // The filesystem to query on cache miss
+  private final org.apache.hadoop.fs.FileSystem mExternalFileSystem;
+  private CacheManager mCacheManager;
+  private org.apache.hadoop.conf.Configuration mHadoopConf;
+  private AlluxioConfiguration mAlluxioConf;
+
+  /**
+   * @param fileSystem File System instance
+   */
+  LocalCachingFileSystem(org.apache.hadoop.fs.FileSystem fileSystem) {
+    mExternalFileSystem = Preconditions.checkNotNull(fileSystem, "filesystem");
+  }
+
+  @Override
+  public synchronized void initialize(URI uri, org.apache.hadoop.conf.Configuration conf)
+      throws IOException {
+    super.initialize(uri, conf);
+    mHadoopConf = conf;
+
+    // Set statistics
+    setConf(conf);
+
+    // Take hadoop configuration to merge to Alluxio configuration
+    Map<String, Object> hadoopConfProperties =
+        HadoopConfigurationUtils.getConfigurationFromHadoop(conf);
+    LOG.info("Creating Alluxio configuration from Hadoop configuration {}", hadoopConfProperties);
+    AlluxioProperties alluxioProps = ConfigurationUtils.defaults();
+    // Merge relevant Hadoop configuration into Alluxio's configuration.
+    alluxioProps.merge(hadoopConfProperties, Source.RUNTIME);
+    // Creating a new instanced configuration from an AlluxioProperties object isn't expensive.
+    mAlluxioConf = new InstancedConfiguration(alluxioProps);
+
+    // Handle metrics
+    Properties metricsProperties = new Properties();
+    for (Map.Entry<String, String> entry : conf) {
+      metricsProperties.setProperty(entry.getKey(), entry.getValue());
+    }
+    MetricsSystem.startSinksFromConfig(new MetricsConfig(metricsProperties));
+
+    mCacheManager = CacheManager.Factory.get(mAlluxioConf);
+  }
+
+  @Override
+  public void close() throws IOException {
+    // super.close should be called first before releasing the resources in this instance, as the
+    // super class may invoke other methods in this class. For example,
+    // org.apache.hadoop.fs.FileSystem.close may check the existence of certain temp files before
+    // closing
+    super.close();
+    mExternalFileSystem.close();
+  }
+
+  /**
+   * @return scheme
+   */
+  // @Override This doesn't exist in Hadoop 1.x, so cannot put {@literal @Override}.
+  public String getScheme() {
+    return mExternalFileSystem.getScheme();
+  }
+
+  @Override
+  public FSDataInputStream open(Path path, int bufferSize) throws IOException {
+    if (mCacheManager == null) {
+      return mExternalFileSystem.open(path, bufferSize);
+    }
+    return new FSDataInputStream(new HdfsFileInputStream(
+        new LocalCacheFileInStream(
+            new AlluxioURI(path.toString()),
+            OpenFilePOptions.getDefaultInstance(),
+            new AlluxioHdfsFileSystem(mExternalFileSystem, mAlluxioConf), mCacheManager),
```

Comment:
Although its a wrapper, do we want to make a new Filesystem every time?

Reply:
created a member variable to save creation overhead